naff/api/gateway/state.py

```diff
@@ -91,13 +96,17 @@ async def stop(self) -> None:
 
     async def _ws_connect(self) -> None:
         """Connect to the Discord Gateway."""
-        log.info("Attempting to initially connect to gateway...")
+        log.info(f"Shard {self.shard_id} is attempting to connect to gateway...")
         try:
             async with GatewayClient(self, (self.shard_id, self.client.total_shards)) as self.gateway:
                 try:
                     await self.gateway.run()
                 finally:
-                    self.client.dispatch(events.Disconnect())
+                    self._shard_ready.clear()
+                    if self.client.total_shards == 1:
+                        self.client.dispatch(events.Disconnect())
+                    else:
+                        self.client.dispatch(events.ShardDisconnect())
```

Comment:
I would expect both to be dispatched
```suggestion
                    if self.client.total_shards == 1:
                        self.client.dispatch(events.Disconnect())

                    self.client.dispatch(events.ShardDisconnect())
```

Reply:
I disagree; disconnect means the entire bot has disconnected, when shard_disconnect is dispatched, the bot could still be connected. I should add a condition to handle all shards being disconnected, however.